larrecodnn/CVN/func/CVNImageUtils.cxx

```diff
@@ -0,0 +1,532 @@
+      windowCharge += wireCharges[nextwire];
+    }
+    if (windowCharge > maxCharge) {
+      maxCharge = windowCharge;
+      firstWire = wire;
+    }
+  }
+  minWire = firstWire;
+  maxWire = firstWire + fNWires - 1;
+
+  std::cout
+    << "Used alternate method to get min and max wires due to vertex determination failure: "
+    << minWire << ", " << maxWire << std::endl;
+}
+
+void lcvn::CVNImageUtils::GetMinMaxTDCs(std::vector<float>& tdcCharges,
+                                        unsigned int& minTDC,
+                                        unsigned int& maxTDC)
+{
+
+  minTDC = 0;
+  maxTDC = fNTDCs;
+
+  for (unsigned int tdc = 0; tdc < tdcCharges.size(); ++tdc) {
+
+    // If we have got to fNWires from the end, the start needs to be this wire
+    if (tdcCharges.size() - tdc == fNTDCs) { break; }
+
+    // For a given tdc, look to see if the next 20 tdcs are empty. If not, this can be out start tdc.
+    int nEmpty = 0;
+    for (unsigned int nextTDC = tdc + 1; nextTDC <= tdc + 20; ++nextTDC) {
+      if (tdcCharges[nextTDC] == 0.0) ++nEmpty;
+    }
+    if (nEmpty < 5) {
+      minTDC = tdc;
+      maxTDC = tdc + fNTDCs - 1;
+      return;
+    }
+  }
+
+  // If we don't find a region where we have fewer than 5 empty tdcs then we just want to select the fNTDCs tdcs containing
+  // the most charge
+  float maxCharge = 0.;
+  unsigned int firstTDC = 0;
+  for (unsigned int tdc = 0; tdc < tdcCharges.size() - fNTDCs; ++tdc) {
+    float windowCharge = 0.;
+    for (unsigned int nexttdc = tdc; nexttdc < tdc + fNTDCs; ++nexttdc) {
+      windowCharge += tdcCharges[nexttdc];
+    }
+    if (windowCharge > maxCharge) {
+      maxCharge = windowCharge;
+      firstTDC = tdc;
+    }
+  }
+  minTDC = firstTDC;
+  maxTDC = firstTDC + fNTDCs - 1;
+
+  std::cout << "Used alternate method to get min and max tdcs due to vertex determination failure: "
+            << minTDC << ", " << maxTDC << std::endl;
+}
+
+void lcvn::CVNImageUtils::ReverseView(std::vector<float>& peVec)
+{
+
+  std::vector<float> vecCopy(peVec.size(), 0.);
+
+  for (unsigned int w = 0; w < fPixelMapWires; ++w) {
+    // Get our new plane number
+    unsigned int newPlane = fPixelMapWires - w - 1;
+
+    for (unsigned int t = 0; t < fPixelMapTDCs; ++t) {
+      float val = peVec[t + fPixelMapTDCs * w];
+      vecCopy[t + fPixelMapTDCs * newPlane] = val;
+    }
+  }
+
+  // Copy the values back into the original vector
+  for (unsigned int e = 0; e < peVec.size(); ++e) {
+    float val = vecCopy[e];
+    peVec[e] = val;
+  }
+}
+
+lcvn::ViewVectorF lcvn::CVNImageUtils::ConvertViewVecToViewVecF(lcvn::ViewVector view)
+{
+
+  lcvn::ViewVectorF newVec;
+  for (size_t w = 0; w < view.size(); ++w) {
+    std::vector<float> thisWire;
+    for (size_t t = 0; t < view[w].size(); ++t) {
+      float chargeSC = static_cast<float>(view[w][t]);
+      thisWire.push_back(chargeSC);
+    }
+    newVec.push_back(thisWire);
+  }
+  return newVec;
+}
```

Comment:
Here (and elsewhere), there are some inefficiencies using vectors:

- The vector passed into the function should be passed by `const` reference; otherwise there's a large expense in copying the vector into the function's stack.
- If you know what the size of the returned vector is going to be whenever the vector is created, then it's best to call `reserve(...)` on that vector to avoid unnecessary reallocations.
- Whenever you have vectors-of-vectors (which is not generally the best idea, but sometimes you're forced into it), make sure you're moving the vectors into the last container instead of copying them.

So the general pattern would look like:
```c++
std::vector<std::vector<float>> toFloats(
  std::vector<std::vector<SomethingElse>> const& view // <= by const reference
)
{
  lcvn::ViewVectorF newVec;
  newVec.reserve(view.size());  // <= reserve
  for (size_t w = 0; w < view.size(); ++w) {
    std::vector<float> thisWire;
    thisWire.reserve(view[w].size());  // <= reserve
    for (size_t t = 0; t < view[w].size(); ++t) {
      thisWire.push_back(view[w][t]);
    }
    newVec.push_back(move(thisWire));   // <= move
  }
  return newVec;
}
```